presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PullConstantsAboveGroupBy.java

```diff
@@ -0,0 +1,142 @@
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.SystemSessionProperties.isOptimizeConstantGroupingKeys;
+import static com.facebook.presto.spi.plan.AggregationNode.singleGroupingSet;
+import static com.facebook.presto.sql.analyzer.ExpressionTreeUtils.isConstant;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.relational.OriginalExpressionUtils.castToExpression;
+import static com.facebook.presto.sql.relational.OriginalExpressionUtils.isExpression;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.function.Function.identity;
+
+/**
+ * Transforms:
+ * <pre>
+ * - GroupBy key1, <const_expr>, key2
+ * </pre>
+ * Into:
+ * <pre>
+ * - Project <const_expr>
+ *    - GroupBy key1, key2
+ * </pre>
+ */
+public class PullConstantsAboveGroupBy
+        implements Rule<AggregationNode>
+{
+    private static final Capture<ProjectNode> CHILD = Capture.newCapture();
+
+    private static final Pattern<AggregationNode> PATTERN =
+            aggregation()
+            .matching(agg -> agg.getGroupingSetCount() == 1)
+            .with(source().matching(project().capturedAs(CHILD)));
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeConstantGroupingKeys(session);
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode parent, Captures captures, Context context)
+    {
+        if (!isEnabled(context.getSession())) {
+            return Result.empty();
+        }
+
+        // for each variable references in grouping keys, check if the CHILD expression defines them as constants
+        ProjectNode child = captures.get(CHILD);
+        List<VariableReferenceExpression> outputVariables = parent.getOutputVariables();
+
+        Map<VariableReferenceExpression, RowExpression> constChildVars = extractConstVars(child, outputVariables);
+
+        List<VariableReferenceExpression> groupingKeys = parent.getGroupingKeys();
+        List<VariableReferenceExpression> newGroupingKeys =
+                groupingKeys.stream()
+                        .filter(key -> !constChildVars.containsKey(key))
+                        .collect(toImmutableList());
+
+        if (constChildVars.isEmpty() || newGroupingKeys.equals(groupingKeys)) {
+            return Result.empty();
+        }
+
+        AggregationNode newAgg = new AggregationNode(
+                parent.getSourceLocation(),
+                parent.getId(),
+                child,
```

Comment:
And you should rewrite the child to drop the constant assignment and outputs

Reply:
this is not always correct, the constant variable may be used as an argument to one of the agg functions, for example 
```
select avg(x) from (select 1 as x) group by x;
```

I think pruning unused columns should be done separately as it may show up in other cases as well

Reply:
You can call PruneUnreferencedOutputs after this  optimizer runs